### PR TITLE
2789 Add ToDevice transform

### LINKED ISSUE #2789: Add `ToDevice` transform to execute transform logic on GPU
**Is your feature request related to a problem? Please describe.**
With `ToDevice` after `ToTensor` or `EnsureType`, we can move the data to GPU and leverage `CacheDataset` to avoid duplicated CPU -> GPU copying in every epoch. And also can support other GPU transforms to accelerate.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #2789 .

### Description
This PR added the `ToDevice` transform.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: tests/test_to_deviced.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+from monai.transforms import ToDeviced
+from tests.utils import skip_if_no_cuda
+
+
+@skip_if_no_cuda
+class TestToDeviced(unittest.TestCase):
+    def test_value(self):
+        device = "cuda:0"
+        data = [{"img": torch.tensor(i)} for i in range(4)]
+        dataset = CacheDataset(data=data, transform=ToDeviced(keys="img", device=device), cache_rate=1.0)
```
[the remark this answers is not in the document]
Comment (dongyang0122) [reply tests/test_to_deviced.py:26]: @Nic-Ma

[the remark this answers is not in the document]
Comment (Nic-Ma) [reply tests/test_to_deviced.py:26]: Hi @dongyang0122 ,

Thanks for your review.
Here to show an important usage: cache data to GPU by `CacheDataset` then access it in DataLoader, as GPU can't work in multi-processing, so used `ThreadDataLoader`.

Thanks.

File: monai/transforms/utility/array.py [hunk 2]
```diff
@@ -1015,3 +1016,25 @@ def _compute(op: Callable, data: np.ndarray):
                 raise ValueError("ops must be key string for predefined operations or callable function.")
 
         return img, meta_data
+
+
+class ToDevice:
```
Comment (wyli) [inline monai/transforms/utility/array.py:1021]: I think this should inherit Transform

Comment (Nic-Ma) [reply monai/transforms/utility/array.py]: Thanks for pointing out the typo, updated in the latest commit.

**Comment by rijobro** [comment]
Hi @Nic-Ma, what's the motivation of this one? I thought we previously said ([link](https://github.com/Project-MONAI/MONAI/pull/1489)) said it wasn't a good idea as the batch wouldn't be contiguously copied to the device?

**Comment by Nic-Ma** [comment]
> Hi @Nic-Ma, what's the motivation of this one? I thought we previously said ([link](https://github.com/Project-MONAI/MONAI/pull/1489)) said it wasn't a good idea as the batch wouldn't be contiguously copied to the device?

Hi @rijobro ,

With this transform, we can cache the data in GPU directly then avoid duplicated copying from CPU to GPU in every epoch, and execute following transforms with GPU tensor to accelerate the training progress.
I tested locally with the spleen segmentation tutorial, it can help improve the training speed by 30%.

Thanks.

**Comment by wyli** [comment]
Thanks for the reminder @rijobro  https://github.com/Project-MONAI/MONAI/pull/1489 Provides more options such as non-blocking... Shall we consider those? Or reopen the PR and work on that one

**Comment by Nic-Ma** [comment]
If you guys think `non_blocking` is necessary, I can also add it.
Thanks for the great discussion!

**Comment by Nic-Ma** [comment]
Hi @wyli @rijobro ,

Do you guys have other concerns or comments here?

Thanks in advance.

**Comment by wyli** [comment]
> If you guys think `non_blocking` is necessary, I can also add it.
> Thanks for the great discussion!

I'm not familiar with this option, @dongyang0122 any idea?

**Comment by dongyang0122** [comment]
> > If you guys think `non_blocking` is necessary, I can also add it.
> > Thanks for the great discussion!
> 
> I'm not familiar with this option, @dongyang0122 any idea?

The transform is helpful and necessary to enable GPU-based data pre-processing or augmentation. The computing efficiency will be greatly improved.

**Comment by wyli** [comment]
thanks @dongyang0122 , the `.to()` API of pytorch has a `non_blocking=False` by default, https://pytorch.org/docs/stable/generated/torch.Tensor.to.html do you have experience with this flag?

**Comment by Nic-Ma** [comment]
Hi @wyli ,

As the doc said: "When non_blocking, tries to convert asynchronously with respect to the host if possible, e.g., converting a CPU Tensor with pinned memory to a CUDA Tensor. "
Maybe let me add a `**kwargs` to this `ToDevice` transform?

Thanks.

**Comment by wyli** [comment]
thanks @Nic-Ma, sounds good

**Review by wyli — APPROVED** [review]

**Pull request merged**